### PR TITLE
Update AWS S3 Registry Backend Topic

### LINKED ISSUE #236: Add information on IAM Policy/Roles to Registry Backend Doc
@jwitz commented on [Fri Dec 18 2020](https://github.com/astronomer/docs-archived/issues/145)

@krisdock commented on [Thu Nov 05 2020](https://github.com/astronomer/issues/issues/2087)

**Describe the document**

there's an "undocumented feature" with our s3 registry backend setup. as seen in [this customer ticket](https://astronomer.zendesk.com/agent/tickets/2921), you can attach the s3 write iam policy to your instance profile and leave the aws credential helm values empty. we have similar instructions for the iam policy and trust relationship setup for forwarding logs to s3 [here](https://www.astronomer.io/docs/enterprise/v0.16/monitor/logs-to-s3/). this allows customers to configure the necessary write access to s3 for their registry backend without having to store aws credentials in plain text in a helm file.

i propose we either call out this functionality as an option, or, providing this is tested and proven to work as stated, we switch to this method and stop instructing customers to use the aws credentials helm values.

**Information**

Check all that apply:
- [x] Was this directly requested by a customer? If so, which one? 
logmein
- [ ] Does this relate to a Cloud use case?
- [x] Does this relate to an Enterprise use case? 

**Organization**

If not, which document should be modified with the missing information?
https://www.astronomer.io/docs/enterprise/v0.16/manage-astronomer/registry-backend/

**User Story**
What user story does this doc solve?
added security

**Screenshots**

**Alternatives**
Are there any alternatives to consider?

**Additional context**

---

@ianstanton commented on [Tue Jan 26 2021](https://github.com/astronomer/docs-archived/issues/145#issuecomment-767815250)

@krisdock Your explanation of this issue is spot on. Thank you for bringing this to light.

> i propose we either call out this functionality as an option, or, providing this is tested and proven to work as stated, we switch to this method and stop instructing customers to use the aws credentials helm values.

I vote that we keep the option to use credentials, but also document instructions for using an IAM role.  

---

@ianstanton commented on [Tue Jan 26 2021](https://github.com/astronomer/docs-archived/issues/145#issuecomment-767815681)

@shmanu017 has told me that Autodesk is also interested in these instructions.

---

@paolaperaza commented on [Fri Jan 29 2021](https://github.com/astronomer/docs-archived/issues/145#issuecomment-770025396)

Similar question from a customer today: https://astronomer.zendesk.com/agent/tickets/3568

---

@shmanu017 commented on [Thu Feb 11 2021](https://github.com/astronomer/docs-archived/issues/145#issuecomment-770048333)

Steps to policy-based access(+ @stevenjs1 )
1. Create an IAM policy [example](https://www.astronomer.io/docs/enterprise/v0.23/manage-astronomer/registry-backend#aws-s3)
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "s3:ListBucket",
                "s3:GetBucketLocation",
                "s3:ListBucketMultipartUploads"
            ],
            "Resource": "arn:aws:s3:::<bucket name>"
        },
        {
            "Effect": "Allow",
            "Action": [
                "s3:PutObject",
                "s3:GetObject",
                "s3:DeleteObject",
                "s3:ListMultipartUploadParts",
                "s3:AbortMultipartUpload"
            ],
            "Resource": "arn:aws:s3:::<bucket name>/*"
        }
    ]
}
```
2. Navigate to EC2 worker node -> Select a worker, Navigate to Security Tab -> Select the IAM role and add custom IAM policy

3. Update the helm values to ([ref](https://docs.docker.com/registry/storage-drivers/s3/))

```
astronomer:
  registry:
    s3:
      enabled: true
      region: us-east-1
      bucket: <bucket_name>

```

---

@stevenjs1 commented on [Tue Mar 09 2021](https://github.com/astronomer/docs-archived/issues/145#issuecomment-794225739)

Looks good to me.

### PR DESCRIPTION
**Pull request by kwcanuck**
Resolves #236

### CODE REVIEW

File: software/registry-backend.md [hunk 1]
```diff
@@ -136,30 +134,40 @@ To grant the registry appropriate push and pull permissions, follow the steps be
...
-4. Push the configuration change to your platform as described in [Apply a config change](apply-platform-config.md).
+  ```yaml
+    astronomer:
+      registry:
+        s3:
+          enabled: true
+          accesskey: my-access-key
+          secretkey: my-secret-key
+          region: us-east-1
+          bucket: <your-bucket-name>
```
Comment (shmanu017) [inline software/registry-backend.md:151]: https://github.com/astronomer/astronomer/pull/1607
The current configuration will have 1 more parameter `regionendpoint`  once this PR is merged.

Comment (kwcanuck) [reply software/registry-backend.md:152]: This change has been added.

File: software/registry-backend.md [hunk 2]
```diff
@@ -136,30 +134,42 @@ To grant the registry appropriate push and pull permissions, follow the steps be
...
+4. Push the configuration change to your platform. See [Apply a config change](apply-platform-config.md).
 
 ### Enable encryption (_Optional_)
 
-To enable encryption, follow the steps below.
-
 1. Create a key in AWS Key Management Service (KMS). During the key creation process you'll be asked to add "key users". Add the user created above as a "key user".
 
-2. Enable encryption by adding the following values to your `config.yaml` file:
+2. Add the following values to your `config.yaml` file to enable encryption:
```
Comment (danielhoherd) [inline software/registry-backend.md:172]: Why did we change this wording? It seems to me like having feature and the fact that it's being toggled at the beginning of the line makes it easier to skim and grok what that line is saying. The new wording requires us to parse that a config file is being modified without first knowing what feature is being modified within the config.

As a dev I would prefer the original wording.

File: software/registry-backend.md [hunk 3]
```diff
@@ -174,13 +184,13 @@ astronomer:
       keyid: my-kms-key-id
 ```
 
-3. Push the configuration change to your platform as described in [Apply a config change](apply-platform-config.md).
+3. Push the configuration change to your platform. See [Apply a config change](apply-platform-config.md).
```
Comment (danielhoherd) [inline software/registry-backend.md:187]: I think we should expand the second sentence here with a qualifier, just to be clear that users do not always need to see that link. EG:
```suggestion
3. Push the configuration change to your platform. See [Apply a config change](apply-platform-config.md) for more details on how this is performed.
```

File: software/registry-backend.md (hunk 2 above)
Comment (kwcanuck) [reply software/registry-backend.md:172]: @danielhoherd this wording was changed to make it consistent with our other documentation.

File: software/registry-backend.md (hunk 3 above)
Comment (kwcanuck) [reply software/registry-backend.md:187]: @danielhoherd this addition is unnecessary as users can infer where the link will take them by the link text.

File: software/registry-backend.md (hunk 2 above)
Comment (danielhoherd) [reply software/registry-backend.md:172]: As an engineer I'd prefer the wording to be easy to grok more than being consistent. Better yet, make all the docs consistent where the important part of the line is at the beginning of the sentence.

As a comparison, we wouldn't say "add the following contents to a bowl and microwave it for 3 minutes to make oatmeal", we would say "to make oatmeal, microwave the following ingredients in a bowl for 3 minutes".

Comment (kwcanuck) [reply software/registry-backend.md:172]: @danielhoherd we typically start our procedure steps with verbs and not imperative statements. This is a hot topic and much debated by technical writers. I would argue that the two statement examples you've provided for making oatmeal work equally well. It's often personal preference that determines what works best. In my 23 years as a professional writer, I've seen companies change from one format to another with little or no concern expressed by users.

**Review by danielhoherd — APPROVED** [review]
Seems OK to me.

**Review by shmanu017 — APPROVED** [review]

**Pull request merged**